### DISCUSSION TITLE
Cannot start dinit as user

### DISCUSSION THREAD

**Discussion by root2185** [davmac314/dinit#189] [Q&A] [answered]
I am on Artix with dinit as init.
```
Kernel: 6.3.1-artix1-1 
Arch: x86_64
Distro: Artix Linux
```
I made the files in my home:
```
cat ~/.config/dinit.d/boot
type = internal
waits-for.d = boot.d
```
If I try to start dinit:
```
dinit
dinit: In multiple cgroups, cannot determine cgroup root path
dinit: Control socket is already active (another instance already running?)
```
I am using root as my user, can I use dinit for user services if I am root?

**Comment by davmac314** [comment, marked as answer]
You can use dinit for user services as root. From the manual page:

```
       -s, --system
              Run as the system service manager.  This is the default if
              invoked as the root user.  This option affects the default
              service definition directory and control socket path.
```
and
```
       -u, --user
              Run as a user service manager.  This is the opposite of
              --system, and is the default if not invoked as the root user.
```

I hope that makes it clear. You need to use `-u` for a user instance if running as root.

**Reply by root2185** [reply]
Nice therefore you use `-u` for `dinitctl` too.

Must log's folders always be created manually before trying to use the service?

Is there a dbus user service that I can use? 
Now I am launching it with the session command, e.g. `exec dbus-launch --exit-with-session bspwm`.

This one https://github.com/Xynonners/dinit-userservd-services/tree/main was causing the system to freeze while cpu usage started to skyrocket.
I tried it again looks like it's causing that problem again.

**Reply by davmac314** [reply]
> Must log's folders always be created manually before trying to use the service?

Yes, dinit won't create the directory for the log file itself.

>Is there a dbus user service that I can use?
> This one https://github.com/Xynonners/dinit-userservd-services/tree/main was causing the system to freeze while cpu usage started to skyrocket.

I've never seen that issue, sorry, I don't know what is causing it.

**Reply by root2185** [reply]
Are the files from there correct?
```
cat ~/.config/dinit.d/dbus.user 
type = process
load-options = sub-vars
logfile = $HOME/.local/share/dinit/dbus.log
command = /etc/dinit.d/user/scripts/dbus-session --print-address 3
smooth-recovery = true
ready-notification = pipefd:3

cat /etc/dinit.d/user/scripts/dbus-session
#!/bin/sh

export DBUS_SESSION_BUS_ADDRESS="unix:path=/run/user/`id -u`/bus"

# export into activation environment to make it visible to other services
dinitctl setenv DBUS_SESSION_BUS_ADDRESS || :

exec dbus-daemon --session --address="$DBUS_SESSION_BUS_ADDRESS" "$@"

```

**Reply by mobin-2008** [reply]
I suggest to run `/etc/dinit.d/user/scripts/dbus-session` by hand in your shell and see what's wrong.

**Reply by root2185** [reply]
```
/etc/dinit.d/user/scripts/dbus-session --print-address 3
dbus-daemon[11759]: Failed to start message bus: Writing to pipe: Invalid argument

```
So I removed `3`:
```
/etc/dinit.d/user/scripts/dbus-session --print-address 
unix:path=/run/user/0/bus,guid=2040b7ed499bb64063ceaaad646bac45
mdbus-daemon[12570]: [session uid=0 pid=12570] Activating service name='org.gtk.vfs.Daemon' requested by ':1.0' (uid=0 pid=13906 comm="mate-system-monitor")
dbus-daemon[12570]: [session uid=0 pid=12570] Successfully activated service 'org.gtk.vfs.Daemon'
dbus-daemon[12570]: [session uid=0 pid=12570] Activating service name='org.gtk.vfs.UDisks2VolumeMonitor' requested by ':1.0' (uid=0 pid=13906 comm="mate-system-monitor")
dbus-daemon[12570]: [session uid=0 pid=12570] Successfully activated service 'org.gtk.vfs.UDisks2VolumeMonitor'
dbus-daemon[12570]: [session uid=0 pid=12570] Activating service name='ca.desrt.dconf' requested by ':1.0' (uid=0 pid=13906 comm="mate-system-monitor")
dbus-daemon[12570]: [session uid=0 pid=12570] Successfully activated service 'ca.desrt.dconf'
```
With just `mate-system-monitor` opened the cpu fan started to spin heavily and I had high usage of cpu.
So I killed the dbus process.
But the fan kept spinning and also the high cpu usage remained.

**Reply by mobin-2008** [reply]
So seems like it's something related to `dbus` and its things rather than `dinit`.

**Reply by root2185** [reply]
Should I open an issue there?

Are the dbus service files good? Do you use those too?

**Reply by davmac314** [reply]
You should probably at least determine which process is using a lot of CPU. It might not be dbus. But it's really outside the scope of what I can provide support for, sorry.

**Reply by root2185** [reply]
That's why I opened the system monitor, but I saw as usually only the `Xorg` process using some.
I am having this problem since I stated using dinit for also user services.
Maybe just removing/readding dbus (`exec dbus-launch --exit-with-session`) to the launcher causes this, so I don't know if it is related to dinit.

I made a service to start dinit as a user instance:
```
cat /etc/dinit.d/dinit.user 
type = process
command = "dinit -u"
smooth-recovery = true
logfile = /var/log/dinit/user-suite.log
waits-for = elogind
waits-for = tmpfs
run-as = root
```
`dinitctl` shows it fine but when I use `dinitctl -u` there is no dinit user instance running.

**Reply by davmac314** [reply]
`"dinit -u"` should not be quoted or it is interpreted as the executable name. The command line should just say:
```
command = dinit -u
```
It makes no sense that `dinitctl` would show the service as running if the command was set like that. Perhaps you edited the service file after already loading it and forgot to reload it via `dinit reload dinit.user`?

**Reply by davmac314** [reply]
In any case if there is some oddity in some service then the things to check and report would be:
- status of the service as reported by `dinitctl status` ...
- whether the process is actually running, whether its pid matches what `status` reported
- anything in dinit's log
- anything in the service's log

**Reply by root2185** [reply]
I tried without quoting it before then that, since I couldn't use `dinitctl -u`.

1. 

```
dctl status dinit.user
Service: dinit.user
    State: STARTED
    Activation: explicitly started
    Process ID: 10384
```
2. Process is running and pid matches.
3. Dinit log? Where is it?
4. Service log:
```
[STOPPD] boot
dinit: In multiple cgroups, cannot determine cgroup root path
dinit: picom: execution failed - setting up standard input/output descriptors: No such file or directory
[FAILED] picom
dinit: pipewire.user: execution failed - setting up standard input/output descriptors: No such file or directory
[FAILED] pipewire.user
[FAILED] wireplumber.user
[FAILED] pipewire-pulse.user
[  OK  ] boot
```
Those are the user services in `~/.config/dinit.d/boot.d/`.

If I try to check something:
```
dinitctl -u list
dinit-client: connecting to socket /run/user/0/dinitctl: No such file or directory
```

**Reply by davmac314** [reply]
> I tried without quoting it before then that, since I couldn't use dinitctl -u

Ok but as I say, with quoting around the whole thing it definitely won't run anything ("dinit -u" is not the name of the executable) and the service would show as failed.

I think the problem you have now is that `dinit` and `dinitctl` are using different locations for the control socket. That's because in user instances the XDG_RUNTIME_DIR environment variable is used to locate the socket. However, XDG_RUNTIME_DIR is associated with your login session (it is set up by elogind) and since you are running dinit (user instance) directly as a service from the system instance, it is not in the login session and does not have XDG_RUNTIME_DIR set, nor will the directory it refers to (`/run/user/0`) exist at that time.

You could use the `-p` argument to dinitctl to specify the control socket location, it should default to `/root/.dinitctl` assuming that the home directory of root is `/root`. Eg:
```
/root/dinitctl -u -p /root/.dinitctl list
```

However you may well run into other issues with this setup, since any services that you run from the user instance of dinit will also not be in the login session. There's a discussion of that here: https://github.com/davmac314/dinit/discussions/183

As noted there, the best solution is probably to start the user instance of dinit via your `.profile` file for example so that it runs once you log in. Another option is to user dinit-userservd/turnstile.

Finally, it looks like some of your services (picom, pipewire) are failing because the log directory specified does not exist when the services are started.

**Reply by root2185** [reply]
I am starting dinit from bspwm autostart and it works fine.
`dinit-userservd` doesn't work.
Does turnstile work on Artix? https://github.com/chimera-linux/turnstile/issues/6

> Finally, it looks like some of your services (picom, pipewire) are failing because the log directory specified does not exist when the services are started.

The log directories exist, when the control socket it's set correctly those services work fine.

**Reply by davmac314** [reply]
> Does turnstile work on Artix?

I have no experience with turnstile, sorry.

> The log directories exist, when the control socket it's set correctly those services work fine.

They may exist now that you are starting dinit from the login session (via bspwm), but at the time dinit was starting earlier, they did not exist. That's why you get that particular message:

> dinit: picom: execution failed - setting up standard input/output descriptors: No such file or directory

**Reply by root2185** [reply]
My CPU usage problem was caused by this app https://github.com/thomas10-10/foo-Wallpaper-Feh-Gif.

> They may exist now that you are starting dinit from the login session (via bspwm), but at the time dinit was starting earlier, they did not exist. That's why you get that particular message:

I deleted the log `/var/log/dinit/user-suite.log` and I started the `dinit.user `service again, the log is as before.

Is suggested to start dbus with dinit? 
Because it doesn't work well as using `exec dbus-launch --exit-with-session` in the launcher.
I can't use `epiphany`, as when I wasn't starting dbus.

**Comment by davmac314** [comment]
If by same as before you mean it still contains:
```
dinit: picom: execution failed - setting up standard input/output descriptors: No such file or directory
dinit: pipewire.user: execution failed - setting up standard input/output descriptors: No such file or directory
```
Then it is these user services which specify an invalid / nonexistent log directory.

> Is suggested to start dbus with dinit?

I have not yet tried to set this up myself, but there are details in this post on the Artix forums:
https://forum.artixlinux.org/index.php/topic,3268.0.html

**Reply by root2185** [reply]
> Then it is these user services which specify an invalid / nonexistent log directory.

Those services work fine when I use `dinit -u` manually.
```
dinitctl -u list
[[+]     ] boot
[{+}     ] picom (pid: 16616)
[{+}     ] pipewire-pulse.user (pid: 706)
[{+}     ] pipewire.user (pid: 705)
[{+}     ] wireplumber.user (pid: 707)
```
They are in `~/.config/dinit.d/`.

**Reply by davmac314** [reply]
Then either the directories exist by the time you start it manually, or possibly the `logfile` setting for those services references an environment variable which is set when you start it manually but isn't otherwise.

**Reply by root2185** [reply]
The dirs existed before I tried to launch the `dinit.user` service.

So maybe this can be useful to understand what it's not working well with that, no idea.
```
cat .config/dinit.d/picom 
type = process
command = picom
logfile = $HOME/.local/share/dinit/picom.log
restart = true

cat /etc/dinit.d/dinit.user 
type = process
command = dinit -u
smooth-recovery = true
logfile = /var/log/dinit/user-suite.log
waits-for = elogind
waits-for = tmpfs
run-as = root
```

**Reply by davmac314** [reply]
`logfile = $HOME/.local/share/dinit/picom.log` <-- there's the problem, it uses an environment variable which isn't set when the service is started. So it ends up trying to put the logfile in `/.local/share/dinit`, which doesn't exist.

The other service with a similar issue was pipewire, so I assume it's probably the same problem.

**Reply by root2185** [reply]
I removed the enviroment variable, no more error about that.
But it's like dinit user instance it is not started:
```
dinitctl status dinit.user
Service: dinit.user
    State: STARTED
    Activation: explicitly started
    Process ID: 10828
```
```
dinitctl -u list
dinit-client: connecting to socket /run/user/0/dinitctl: No such file or directory
```

**Reply by davmac314** [reply]
See above:
> I think the problem you have now is that dinit and dinitctl are using different locations for the control socket. That's because in user instances the XDG_RUNTIME_DIR environment variable is used to locate the socket. However, XDG_RUNTIME_DIR is associated with your login session (it is set up by elogind) and since you are running dinit (user instance) directly as a service from the system instance, it is not in the login session and does not have XDG_RUNTIME_DIR set, nor will the directory it refers to (/run/user/0) exist at that time.

etc.

**Reply by root2185** [reply]
Right, should I open an issue about that?

Running `dinit -u` in the autostart looks good.

**Reply by davmac314** [reply]
> Right, should I open an issue about that?

As far as I'm concerned it's working as intended. The "fix" is not to run your user instance of dinit directly from the system instance, instead, make sure it's run inside the login session.

> Running dinit -u in the autostart looks good.

Not sure what you mean by the autostart but if it works, then great.

**Reply by root2185** [reply]
> Not sure what you mean by the autostart but if it works, then great.

I mean the `~/.config/bspwm/bspwmrc` file, it is executed when the session starts.
